quarkus.ls/com.redhat.quarkus.ls/src/test/java/com/redhat/quarkus/services/ApplicationPropertiesRequiredCodeActionTest.java

```diff
@@ -0,0 +1,209 @@
+		d.add(d1);
+		d.add(d2);
+
+		testDiagnosticsFor(value, projectInfo, validationSettings, d1, d2);
+		testCodeActionsFor(value, d, d.get(0).getRange(), projectInfo,
+				ca("Add all missing required properties?", te(1, 37, 1, 37, "\nquarkus.required.property=\nquarkus.second.required.property="), d));
+	}
+
+	@Test
+	public void testMissingPropertyTrailingWhitespaceCodeAction() throws BadLocationException {
+
+		String value = "quarkus.optional.property=hello\n" + //
+			"quarkus.third.required.property=hello\n      \n     \n     ";
+
+		Diagnostic d1 = d(0, 0, 4, 5, "Missing required property 'quarkus.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+		Diagnostic d2 = d(0, 0, 4, 5, "Missing required property 'quarkus.second.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+
+		List<Diagnostic> d = new ArrayList<>();
+		d.add(d1);
+		d.add(d2);
+
+		testDiagnosticsFor(value, projectInfo, validationSettings, d1, d2);
+		testCodeActionsFor(value, d, d.get(0).getRange(), projectInfo,
+				ca("Add all missing required properties?", te(1, 37, 1, 37, "\nquarkus.required.property=\nquarkus.second.required.property="), d1, d2));
+	}
+
+	@Test
+	public void testMissingPropertyWhitepspaceCodeAction() throws BadLocationException {
+
+		String value = "     \n      \n     \n      \n     \n     ";
+
+
+		Diagnostic d1 = d(0, 0, 5, 5, "Missing required property 'quarkus.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+		Diagnostic d2 = d(0, 0, 5, 5, "Missing required property 'quarkus.second.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+		Diagnostic d3 = d(0, 0, 5, 5, "Missing required property 'quarkus.third.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+
+		List<Diagnostic> d = new ArrayList<>();
+		d.add(d1);
+		d.add(d2);
+		d.add(d3);
+		testDiagnosticsFor(value, projectInfo, validationSettings, d1, d2, d3);
+		testCodeActionsFor(value, d, d.get(0).getRange(), projectInfo,
+				ca("Add all missing required properties?", te(0, 0, 0, 0, "quarkus.required.property=\nquarkus.second.required.property=\nquarkus.third.required.property="), d1, d2, d3));
+	}
+
+	@Test
+	public void testMissingPropertyEmptyFileCodeAction() throws BadLocationException {
+
+		String value = "";
+
+		Diagnostic d1 = d(0, 0, 0, 0, "Missing required property 'quarkus.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+		Diagnostic d2 = d(0, 0, 0, 0, "Missing required property 'quarkus.second.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+		Diagnostic d3 = d(0, 0, 0, 0, "Missing required property 'quarkus.third.required.property'", DiagnosticSeverity.Warning,
+				ValidationType.required);
+
+		List<Diagnostic> d = new ArrayList<>();
+		d.add(d1);
+		d.add(d2);
+		d.add(d3);
+		
+		testDiagnosticsFor(value, projectInfo, validationSettings, d1, d2, d3);
```

Comment:
In case of empty file, the line separator comes from the system line separator. This test doesn't work on Windows OS. Replace with:

```java
String lineSeparator = System.lineSeparator();
testDiagnosticsFor(value, projectInfo, validationSettings, d1, d2, d3);
testCodeActionsFor(value, d, d.get(0).getRange(), projectInfo,
		ca("Add all missing required properties?", te(0, 0, 0, 0, "quarkus.required.property=" + //
				lineSeparator + //
				"quarkus.second.required.property=" + //
				lineSeparator + //
				"quarkus.third.required.property="), d1, d2, d3));
```